### ISSUE TITLE
systemd-coredump does not properly symbolize stack traces

### ISSUE THREAD

**Issue by daandemeyer** [systemd/systemd#35316]
To Reproduce:

```
mkosi -f qemu
[root@archlinux ~]# systemctl kill -s ABRT systemd-networkd
[root@archlinux ~]# coredumpctl info
           PID: 827 (systemd-network)
           UID: 976 (systemd-network)
           GID: 976 (systemd-network)
        Signal: 6 (ABRT)
     Timestamp: Fri 2024-11-22 23:36:53 CET (16min ago)
  Command Line: /usr/lib/systemd/systemd-networkd
    Executable: /usr/lib/systemd/systemd-networkd
 Control Group: /system.slice/systemd-networkd.service
          Unit: systemd-networkd.service
         Slice: system.slice
       Boot ID: e48e7a49b6de4af08f2380ba0db777d1
    Machine ID: be8967b7f49a4eb6ae953b68214f2481
      Hostname: archlinux
       Storage: journal
       Message: Process 827 (systemd-network) of user 976 dumped core.
                
                Stack trace of thread 827:
                #0  0x00007125e6d28877 epoll_wait (libc.so.6 + 0x119877)
                #1  0x00007125e7208c96 n/a (/usr/lib/systemd/libsystemd-shared-257rc2.so (deleted) + 0x408c96)
                #2  0x00007125e7208e54 n/a (/usr/lib/systemd/libsystemd-shared-257rc2.so (deleted) + 0x408e54)
                #3  0x00007125e72095fb n/a (/usr/lib/systemd/libsystemd-shared-257rc2.so (deleted) + 0x4095fb)
                #4  0x00007125e720a000 n/a (/usr/lib/systemd/libsystemd-shared-257rc2.so (deleted) + 0x40a000)
                #5  0x00007125e720a280 n/a (/usr/lib/systemd/libsystemd-shared-257rc2.so (deleted) + 0x40a280)
                #6  0x0000619b40930308 n/a (/usr/lib/systemd/systemd-networkd (deleted) + 0x12308)
                #7  0x0000619b40930533 n/a (/usr/lib/systemd/systemd-networkd (deleted) + 0x12533)
                #8  0x00007125e6c34e08 n/a (libc.so.6 + 0x25e08)
                #9  0x00007125e6c34ecc __libc_start_main (libc.so.6 + 0x25ecc)
                #10 0x0000619b4092f045 n/a (/usr/lib/systemd/systemd-networkd (deleted) + 0x11045)
                ELF object binary architecture: AMD x86-64

```

The stack trace shown will have symbol addresses and not file:line information. `readelf --section-headers /usr/lib/systemd/systemd-networkd` will show that the networkd binary has all the necessary debug sections.

**Comment by DaanDeMeyer** [comment]
cc @d-e-s-o I didn't check with you beforehand but I figured this might be a good use case for blazesym

**Comment by poettering** [comment]
where does that get its data from? minidebugsymbols?

**Comment by poettering** [comment]
afaics blazesym wouldn't help us: line numbers it would get from full dwarf, but libelfutils already gets us that too, if you install full debuginfo.

Or what am i missing.

**Comment by DaanDeMeyer** [comment]
> afaics blazesym wouldn't help us: line numbers it would get from full dwarf, but libelfutils already gets us that too, if you install full debuginfo.
> 
> Or what am i missing.

If I generate a coredump of systemd-networkd from within a mkosi VM, I don't get line numbers in the stacktrace at all, just addresses. In this case every systemd binary and library has full debuginfo.

**Comment by DaanDeMeyer** [comment]
Repurposed into a bug report

**Comment by d-e-s-o** [comment]
I'd be happy to work with you to integrate `blazesym`, if there is interest. I agree with Lennart that it won't provide much value-add here, though, assuming you are already consulting DWARF information. That being said, if you want to check how `blazesym` would symbolize addresses, you can try using [`blazecli`](https://github.com/libbpf/blazesym/tree/main/cli) on the coredump. That could also potentially help narrow down the issue. Statically linked binaries are available [here](https://github.com/libbpf/blazesym/releases/tag/cli-v0.1.7) (they currently include debug information themselves, that's why they are rather large; the library itself is kept slim) in case you don't have a Rust toolchain handy.

**Comment by msekletar** [comment]
@DaanDeMeyer I am not sure but this looks a lot like this https://sourceware.org/bugzilla/show_bug.cgi?id=32713

tldr; there was a kernel change that broke elfutils and is now "fixed" by https://github.com/torvalds/linux/commit/39ec9eaaa165d297d008d1fa385748430bd18e4d.

**Comment by msekletar** [comment]
It seems that, at least on Fedora Rawhide, the backtraces are already properly symbolized.

```
Mon 2026-02-23 09:52:05 EST 1118 192 192 SIGSEGV present  /usr/lib/systemd/systemd-networkd 188.7K
[root@vm-10-0-184-222 ~]# coredumpctl info
           PID: 1118 (systemd-network)
           UID: 192 (systemd-network)
           GID: 192 (systemd-network)
        Signal: 11 (SEGV)
     Timestamp: Mon 2026-02-23 09:52:05 EST (28s ago)
  Command Line: /usr/lib/systemd/systemd-networkd
    Executable: /usr/lib/systemd/systemd-networkd
 Control Group: /system.slice/systemd-networkd.service
          Unit: systemd-networkd.service
         Slice: system.slice
       Boot ID: ea802411350d42538c3efdbaaaae2f61
    Machine ID: 23112c7c3df34417ad0b7d0b5a333494
      Hostname: vm-10-0-184-222.hosted.upshift.rdu2.redhat.com
       Storage: /var/lib/systemd/coredump/core.systemd-network.192.ea802411350d42538c3efdbaaaae2f61.1118.1771858325000000.zst (present)
  Size on Disk: 188.7K
       Message: Process 1118 (systemd-network) of user 192 dumped core.

                Module libblkid.so.1 from rpm util-linux-2.41.3-12.fc44.x86_64
                Module libmount.so.1 from rpm util-linux-2.41.3-12.fc44.x86_64
                Module libpcre2-8.so.0 from rpm pcre2-10.47-1.fc44.1.x86_64
                Module libselinux.so.1 from rpm libselinux-3.10-1.fc44.x86_64
                Module libz.so.1 from rpm zlib-ng-2.3.3-2.fc45.x86_64
                Module ld-linux-x86-64.so.2 from rpm glibc-2.43.9000-2.fc45.x86_64
                Module libm.so.6 from rpm glibc-2.43.9000-2.fc45.x86_64
                Module libcrypto.so.3 from rpm openssl-3.5.5-1.fc44.x86_64
                Module libc.so.6 from rpm glibc-2.43.9000-2.fc45.x86_64
                Stack trace of thread 1118:
                #0  0x00007f818ea7524e __internal_syscall_cancel (libc.so.6 + 0x7024e)
                #1  0x00007f818ea75274 __syscall_cancel (libc.so.6 + 0x70274)
                #2  0x00007f818eafcd75 epoll_wait (libc.so.6 + 0xf7d75)
                #3  0x00007f818efe4fd4 epoll_wait_usec (libsystemd-shared-260.so + 0x3e4fd4)
                #4  0x00007f818efe5169 process_epoll (libsystemd-shared-260.so + 0x3e5169)
                #5  0x00007f818efe574c sd_event_wait (libsystemd-shared-260.so + 0x3e574c)
                #6  0x00007f818efe60ae sd_event_run (libsystemd-shared-260.so + 0x3e60ae)
                #7  0x00007f818efe62c2 sd_event_loop (libsystemd-shared-260.so + 0x3e62c2)
                #8  0x00000000004135d9 run (/usr/lib/systemd/systemd-networkd + 0x135d9)
                #9  0x0000000000413757 main (/usr/lib/systemd/systemd-networkd + 0x13757)
                #10 0x00007f818ea08681 __libc_start_call_main (libc.so.6 + 0x3681)
                #11 0x00007f818ea08798 __libc_start_main@@GLIBC_2.34 (libc.so.6 + 0x3798)
                #12 0x0000000000412115 _start (/usr/lib/systemd/systemd-networkd + 0x12115)
                ELF object binary architecture: AMD x86-64
```

**Comment by daandemeyer** [comment]
@msekletar You're supposed to get file + line information when debug symbols are installed.

**Comment by msekletar** [comment]
I don't have the debug symbols installed separately, above was tested on local build. Frankly, I am not sure now whatever I have ever seen backtrace with file+line in coredumpctl info output. In any case at least symbol names are now shown instead of n/a.

**Comment by daandemeyer** [comment]
Hmm maybe I remembered wrong, indeed `frame_callback()` in elf-util.c doesn't show anything like that.

Kinda useful to have in the backtrace though
